clap_builder/src/builder/arg.rs

```diff
@@ -4363,8 +4363,20 @@ impl Arg {
         }
 
         debug_assert!(self.is_takes_value_set());
+
+        let is_required = |n| match self.is_positional() {
+            true => required && (num_vals.min_values() != 0),
+            false => {
+                // If all values are optional, the [] get rendered by the caller:
+                //    --foo[=<bar>]
+                // In this case, treat as required.
+                let required = self.get_min_vals() == 0;
+                required || (n < num_vals.min_values())
+            }
+        };
```

Comment:
nit: personally, I would find it more readable and easier to refactor for this to be code within the for-loop with `is_required` being a variable, rather than a function.

Reply:
Happy to move that back in to an if statement within the loop, I moved it out because of https://github.com/clap-rs/clap/pull/4910#discussion_r1198117667.

Reply:
To be clear, I'm not asking to revert back to a complex if condition but to basically take this logic as structured and just switch it from a closure to assigning to a variable within the loop.

Reply:
Done.